simulation.js

```diff
@@ -1,11 +1,127 @@
+    await dbSetup(database);
+    const workflow = await database.collection("Workflow").find().toArray();
+    const printJob = await database.collection("PrintJob").findOne();
+	simulate(printJob, workflow[0], database);
+}
+
+
 /**
  * Simulate a PrintJob through the given Workflow
  * @param {*} printJob 
  * @param {*} workflow 
- * @returns {*} the SimulationReport as a JSON 
+ * @param {*} database
+ * @returns {*} the ID of the Simulation Report
  */
-async function simulate(printJob, workflow){
-    return 
+
+async function simulate(printJob, workflow, database){
+	//Format workflowSteps correctly for traverseGraph
+	let workflowSteps = {}
+	let temp = workflow.WorkflowSteps;
+	for(let i = 0; i < temp.length; i++){
+		workflowSteps[temp[i]] = {
+			func: (await database.collection("WorkflowStep").findOne({_id: temp[i]})).Title,
+			prev: i == 0 ? [] : [temp[i-1]],
+			next: i == temp.length-1 ? [] : [temp[i+1]]
+		};
+	}
+
+	const results = await traverseGraph(printJob, workflowSteps, 
+		Object.keys(workflowSteps)[0], 
+		Object.fromEntries(Object.keys(workflowSteps).map((k) => [k, false])), 	
+	);
+
+	//Find times for report
+	//TODO: This could be done in a single loop
+	const totalTime = await Math.max(...Object.keys(results).map((k) => 
+		results[k].cumulative));
+
+	const rastTime = await results[Object.keys(results).filter((k) => 
+		results[k].stepName === "Rasterization")].stepTime;
+
+	return await newSimulationReport(database, printJob._id, workflow._id, totalTime, rastTime);
 }
 
+
+async function traverseGraph(printJob, workflowSteps, step, visited, mutex=new Mutex(), results={}){
+	if(await isVisited(visited, step, mutex)) return;
+	await Promise.all(workflowSteps[step].prev.map((k) => 
+		traverseGraph(printJob, workflowSteps, k, visited, mutex, results)));
+	const simulatedTime =  await simulateStep(printJob, workflowSteps, step);
+	results[step] = {stepName: workflowSteps[step].func, stepTime: simulatedTime, cumulative: simulatedTime};
+	results[step].cumulative += await Math.max(workflowSteps[step].prev.map((k) =>
+		results[k].cumulative));
+	await Promise.all(workflowSteps[step].next.map((k) => 
+		traverseGraph(printJob, workflowSteps, k, visited, mutex, results)));
+	return results;
+}
+
+
+async function isVisited(visited, check, mutex){
+	return await mutex.runExclusive(() => {
+		let output = false
+        if (visited[check]) output = true;
+        else visited[check] = true;
+        return output;
+    });
+}
+
+
+async function simulateStep(printJob, workflowSteps, step){
+	// await new Promise(resolve => setTimeout(resolve, Math.random()*1000)); //TODO: Remove
+	const funcs = {
+		"Preflight": preflight,
+		"Metrics": metrics,
+		"Rasterization": rasterization,
+		"Printing": printing,
+		"Cutting": cutting,
+		"Laminating": laminating,
+	}
+	return await funcs[workflowSteps[step].func](printJob)
+}
+
+
+async function preflight(printJob){
```

Comment:
We should not hard code functions for every single possible step because there could be more that the user can create. `WorkflowStep` should contain how much time it takes (see `TimePerPage`), and that is what should be used. So, `WorkflowStep.TimePerPage *  printJob.PageCount`.